tests/drivers/uart/uart_pm/src/main.c

```diff
@@ -0,0 +1,265 @@
+
+		err = memcmp(txbuf, rxbuf, sizeof(txbuf));
+		zassert_equal(err, 0, "Unexpected err: %d", err);
+	}
+
+	zassert_true(tx_done, NULL);
+
+	return true;
+}
+
+static void communication_verify(const struct device *dev, bool active)
+{
+	bool is_async = async_verify(dev, active);
+
+	polling_verify(dev, is_async, active);
+}
+
+#define state_verify(dev, exp_state) do {\
+	uint32_t power_state; \
+	int err = pm_device_state_get(dev, &power_state); \
+	zassert_equal(err, 0, "Unexpected err: %d", err); \
+	zassert_equal(power_state, exp_state, NULL); \
+} while (0)
+
+static void pm_callback(const struct device *dev,
+			int status, uint32_t *state, void *arg)
+{
+	zassert_equal(dev, exp_dev, NULL);
+	zassert_equal(status, 0, NULL);
+	zassert_equal(*state, exp_state, NULL);
+	zassert_equal(arg, exp_arg, NULL);
+	callback_cnt++;
+}
+
+static void state_set(const struct device *dev, uint32_t state, int exp_err, bool cb)
+{
+	int err;
+	uint32_t prev_state;
+
+	err = pm_device_state_get(dev, &prev_state);
+	zassert_equal(err, 0, "Unexpected err: %d", err);
+
+	if (cb) {
+		callback_cnt = 0;
+		exp_dev = dev;
+		exp_arg = &state;
+		exp_state = state;
+
+		err = pm_device_state_set(dev, state, pm_callback, exp_arg);
+		zassert_equal(err, exp_err, "Unexpected err: %d", err);
+		zassert_equal(callback_cnt, 1, NULL);
+	} else {
+		err = pm_device_state_set(dev, state, NULL, NULL);
+		zassert_equal(err, exp_err, "Unexpected err: %d", err);
+	}
+
+	state_verify(dev, err == 0 ? state : prev_state);
+}
+
+static void test_uart_pm_in_idle(void)
+{
+	const struct device *dev;
+
+	dev = device_get_binding(UART_DEVICE_NAME);
+	zassert_true(dev != NULL, NULL);
+
+	state_verify(dev, PM_DEVICE_STATE_ACTIVE);
+	communication_verify(dev, true);
+
+	state_set(dev, PM_DEVICE_STATE_LOW_POWER, 0, false);
+	communication_verify(dev, false);
+
+	state_set(dev, PM_DEVICE_STATE_ACTIVE, 0, false);
+	communication_verify(dev, true);
+
+	state_set(dev, PM_DEVICE_STATE_LOW_POWER, 0, true);
+	communication_verify(dev, false);
+
+	state_set(dev, PM_DEVICE_STATE_ACTIVE, 0, true);
+	communication_verify(dev, true);
+}
+
+static void test_uart_pm_poll_tx(void)
+{
+	const struct device *dev;
+
+	dev = device_get_binding(UART_DEVICE_NAME);
+	zassert_true(dev != NULL, NULL);
+
+	communication_verify(dev, true);
+
+	uart_poll_out(dev, 'a');
+	state_set(dev, PM_DEVICE_STATE_LOW_POWER, 0, false);
+
+	communication_verify(dev, false);
+
+	state_set(dev, PM_DEVICE_STATE_ACTIVE, 0, false);
+
+	communication_verify(dev, true);
+
+	/* Now same thing but with callback */
+	uart_poll_out(dev, 'a');
+	state_set(dev, PM_DEVICE_STATE_LOW_POWER, 0, true);
+
+	communication_verify(dev, false);
+
+	state_set(dev, PM_DEVICE_STATE_ACTIVE, 0, true);
+
+	communication_verify(dev, true);
+}
+
+static void timeout(struct k_timer *timer)
+{
+	const struct device *uart = k_timer_user_data_get(timer);
+
+	state_set(uart, PM_DEVICE_STATE_LOW_POWER, 0, false);
+}
+
+K_TIMER_DEFINE(pm_timer, timeout, NULL);
```

Comment:
This should be `static`, or perhaps it could be defined inside the only function that uses it.